### DISCUSSION TITLE
error "Maximum call stack size exceeded" when extending colors using theme function

### DISCUSSION THREAD

**Discussion by usirg** [tailwindlabs/tailwindcss#5246] [Help] [answered]
Objective:
* redefine the gray color palette 
* add a new key to gray based on the newly set palette

I just created a fresh tailwind project and modified tailwind.config.js like so:
```javascript
const colors = require('tailwindcss/colors')

module.exports = {
  purge: [],
  darkMode: false, // or 'media' or 'class'
  theme: {
    extend: {
      colors: theme => ({
        gray: colors.warmGray,
        gray: {
          'darkest': theme('colors.gray.900'),
        },
      }),  
    },
  },
  variants: {
    extend: {},
  },
  plugins: [],
}
```

I'm running the following command:
`npx tailwindcss -v -i ./src/global.css -o ./dist/tailwind.css`

And I'm getting:
`(node:13434) UnhandledPromiseRejectionWarning: RangeError: Maximum call stack size exceeded
    at colors (/home/markus/tailwind-test/tailwind.config.js:8:25)
    at value (/home/markus/tailwind-test/node_modules/tailwindcss/lib/util/resolveConfig.js:60:53)
    at /home/markus/tailwind-test/node_modules/tailwindcss/lib/util/resolveConfig.js:119:109
    at Array.map (<anonymous>)
    at /home/markus/tailwind-test/node_modules/tailwindcss/lib/util/resolveConfig.js:119:100
    at resolvePath (/home/markus/tailwind-test/node_modules/tailwindcss/lib/util/resolveConfig.js:131:45)
    at colors (/home/markus/tailwind-test/tailwind.config.js:11:22)
    at value (/home/markus/tailwind-test/node_modules/tailwindcss/lib/util/resolveConfig.js:60:53)
    at /home/markus/tailwind-test/node_modules/tailwindcss/lib/util/resolveConfig.js:119:109
    at Array.map (<anonymous>)`

observations:
* the config with the arrow function works as long as I don't call the theme() function inside

Any suggestions?

**Comment by aggreggator** [comment, marked as answer]
There is no need to use the theme()

Does this work for you?
```js
const colors = require('tailwindcss/colors')

module.exports = {
  purge: [],
  darkMode: false, // or 'media' or 'class'
  theme: {
    extend: {
      colors: {
        gray: {
          ...colors.warmGray,
          darkest: colors.warmGray[900],
        },
      },
    },
  },
  variants: {
    extend: {},
  },
  plugins: [],
}

```

**Reply by usirg** [reply]
Thanks. The above does not result in an error.

But it is not exactly what I'm looking for. In case I would like to change the gray palette to e.g. blueGray in the future, I would need to modify the definition of darkest as well and that's what I would like to avoid. Obviously it would be easy to do now. But that changes with time and a more complex config.

The purpose of theme() as per the docs is to enable us to refer to parts of our own config. What I don't know is how to do that in this case (or whether it's possible).

**Reply by aggreggator** [reply]
Using theme() in theme.extend creates a recursion, so the stack exceeded.

My next best guess:

```js
const colors = require('tailwindcss/colors')

const gray = colors.warmGray
// const gray = colors.blueGray 

module.exports = {
  purge: [],
  darkMode: false, // or 'media' or 'class'
  theme: {
    extend: {
      colors: {
        gray: {
          ...gray,
          darkest: gray[900],
        },
      },
    },
  },
  variants: {
    extend: {},
  },
  plugins: [],
}

```

**Reply by usirg** [reply]
Many thanks, I tried and that works.

Note the use of the constant gray in both the override and extend sections below:

```javascript
const colors = require('tailwindcss/colors');
const gray = colors.warmGray;

module.exports = {
  purge: [],
  darkMode: class,
  theme: {
    colors: {
      transparent: 'transparent',
      current: 'currentColor',
      black: colors.black,
      white: colors.white,
      gray: gray,
      indigo: colors.indigo,
      red: colors.rose,
      yellow: colors.amber,
      green: colors.green,
      blue: colors.blue,
      purple: colors.purple,
      pink: colors.pink,      
    },
    extend: theme => ({
      colors: {
        gray: {
          'darkest': gray[900],
        },
      },
    }),
  },
  variants: {
    extend: {},
  },
  plugins: [
    require('@tailwindcss/typography'),
  ],
}

```
